File: bindings/php/src/ProvenError.php
<?php

// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Shared exception class for the proven-servers PHP bindings.
//
// Maps the slot-based context pool error pattern used by every Zig FFI
// implementation to PHP exceptions.

declare(strict_types=1);

namespace ProvenServers;

/**
 * Error codes matching the proven-servers Zig FFI status conventions.
 *
 * Every Zig FFI function returns a u8 status:
 *   0 = success
 *   1 = invalid state (wrong lifecycle phase)
 *   2 = validation failed (bad input)
 * Slot-creating functions return c_int: -1 = pool exhausted.
 */
enum ErrorCode: int
{
    /** Context pool exhausted (64-slot limit). */
    case PoolExhausted = -1;
    /** Invalid or inactive context slot. */
    case InvalidSlot = -2;
    /** Operation rejected: wrong lifecycle state. */
    case InvalidState = 1;
    /** Input validation failed. */
    case ValidationFailed = 2;
    /** Parameter value outside valid ABI tag range. */
    case InvalidParameter = 3;
    /** Fixed-size buffer or array capacity exceeded. */
    case CapacityExceeded = 4;
    /** Unknown FFI error. */
    case Unknown = 255;

    /**
     * Get the default human-readable message for this error code.
     */
    public function defaultMessage(): string
    {
        return match ($this) {
            self::PoolExhausted   => 'context pool exhausted (64-slot limit)',
            self::InvalidSlot     => 'invalid or inactive context slot',
            self::InvalidState    => 'operation rejected: wrong lifecycle state',
            self::ValidationFailed => 'input validation failed',
            self::InvalidParameter => 'parameter value outside valid ABI tag range',
            self::CapacityExceeded => 'fixed-size buffer or array capacity exceeded',
            self::Unknown          => 'unknown FFI error',
        };
    }
}

/**
 * Exception thrown by proven-servers FFI wrapper methods.
 */
class ProvenError extends \RuntimeException
{
    /**
     * @param ErrorCode $errorCode The categorised error code.
     * @param int       $rawCode   The raw integer returned by the FFI function.
     * @param string    $message   Optional human-readable message override.
     */
    public function __construct(
        public readonly ErrorCode $errorCode,
        public readonly int $rawCode = 0,
        string $message = '',
    ) {
        $msg = $message !== '' ? $message : $errorCode->defaultMessage();
        parent::__construct($msg, $errorCode->value);
    }

    /**
     * Interpret a slot-returning FFI call (c_int).
     * Returns the slot index for non-negative values.
     *
     * @param int $raw The raw c_int returned by the FFI create function.
     * @return int The valid slot index.
     * @throws self If no free slot is available.
     */
    public static function checkSlot(int $raw): int
    {
        if ($raw >= 0) {
            return $raw;
        }
        throw new self(ErrorCode::PoolExhausted, $raw);
    }

    /**
     * Interpret a status-returning FFI call (u8).
     * 0 = success, 1 = invalid state, 2 = validation failed.
     *
     * @param int $raw The raw u8 status returned by the FFI function.
     * @throws self If the status indicates failure.
     */
    public static function checkStatus(int $raw): void
    {
        if ($raw === 0) {
            return;
        }

        $code = match ($raw) {
            1 => ErrorCode::InvalidState,
            2 => ErrorCode::ValidationFailed,
            3 => ErrorCode::InvalidParameter,
            4 => ErrorCode::CapacityExceeded,
            default => ErrorCode::Unknown,
        };

        throw new self($code, $raw);
    }
}

File: bindings/ocaml/lib/proven_error.ml
(* SPDX-License-Identifier: MPL-2.0 *)
(* Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk> *)

(** Shared error types for all proven-servers FFI operations.

    Every protocol FFI uses the same slot-based context pool pattern with
    [int] return values (-1 = no slot, 0/1 = success/failure). This module
    maps those patterns to a descriptive OCaml variant type. *)

(** Unified error type for all proven-servers FFI operations. *)
type t =
  | Pool_exhausted
    (** No free context slots available in the pool (64-slot limit). *)
  | Invalid_slot
    (** The slot index is invalid or the context is not active. *)
  | Invalid_state
    (** The operation was rejected because the context is in the wrong
        lifecycle state for the requested transition. *)
  | Invalid_parameter
    (** A parameter value is outside the valid ABI tag range. *)
  | Capacity_exceeded
    (** The operation would exceed a fixed-size buffer or array limit. *)
  | Validation_failed
    (** Input validation failed (e.g. path traversal attack). *)
  | Unknown_error of int
    (** The FFI returned an unexpected or undocumented error code. *)

(** Convert a proven error to a human-readable string. *)
let to_string = function
  | Pool_exhausted -> "context pool exhausted (64-slot limit)"
  | Invalid_slot -> "invalid or inactive context slot"
  | Invalid_state -> "operation rejected: wrong lifecycle state"
  | Invalid_parameter -> "parameter value outside valid ABI tag range"
  | Capacity_exceeded -> "fixed-size buffer or array capacity exceeded"
  | Validation_failed -> "input validation failed"
  | Unknown_error code -> Printf.sprintf "unknown FFI error (code %d)" code

(** Interpret a slot-returning FFI call.

    Returns [Ok slot] for non-negative values, [Error Pool_exhausted] for
    negative values (typically -1). *)
let from_slot raw =
  if raw >= 0 then Ok raw
  else Error Pool_exhausted

(** Interpret a status-returning FFI call (0 = success, non-zero = error).

    - 0 -> [Ok ()]
    - 1 -> [Error Invalid_state]
    - 2 -> [Error Validation_failed]
    - other -> [Error (Unknown_error code)] *)
let from_status raw =
  match raw with
  | 0 -> Ok ()
  | 1 -> Error Invalid_state
  | 2 -> Error Validation_failed
  | n -> Error (Unknown_error n)
